src/evolution_vaccination.m

```diff
@@ -3,36 +3,64 @@ function evolution_vaccination()
     
     n = size(system.reward,1);
     
-    %r_vacc == reward for vaccination
+    % r_vacc == reward for vaccination
     r_vacc = -4;
     
-    
+    % Filter for the convolution
     filter = ones(3);
+    
+    % Number of nearest neighbors
     number_of_neighbors = conv2(ones(n),filter,'same')-1;
-    rewards_of_neighbors = conv2(system.reward,filter,'same')-system.reward;
-    rewards_of_neighbors = rewards_of_neighbors ./ number_of_neighbors;
     
-    proba_change_state = rand(n);
+    % Number of NOT vaccinated nearest neighbors
+    number_of_neighbors_NV = conv2(~system.vaccinated,filter,'same')-(~system.vaccinated);
+    % Mean reward of NOT vaccinated neighbors
+    rewards_of_neighbors_NV = conv2(system.reward.*(~system.vaccinated),filter,'same')-system.reward.*(~system.vaccinated);
+    rewards_of_neighbors_NV = rewards_of_neighbors_NV ./ number_of_neighbors_NV;
+    rewards_of_neighbors_NV (isnan(rewards_of_neighbors_NV)) = 0;
     
-    proba_vaccination = vaccination_probability_1(system.reward-rewards_of_neighbors);
+    % Number of vaccinated nearest neighbors
+    number_of_neighbors_V = conv2(system.vaccinated,filter,'same')-(system.vaccinated);
+    % Mean reward of vaccinated neighbors
+    rewards_of_neighbors_V = conv2(system.reward.*system.vaccinated,filter,'same')-system.reward.*system.vaccinated;
+    rewards_of_neighbors_V = rewards_of_neighbors_V ./ number_of_neighbors_V;
+    rewards_of_neighbors_V (isnan(rewards_of_neighbors_V)) = 0;
     
-    indices_not_vacc = (proba_vaccination>proba_change_state);
+    % Number of infected nearest neighbors
+    indices_of_I = system.state=="I";
+    number_of_neighbors_I = conv2(indices_of_I,filter,'same')-(indices_of_I);    
+    prop_of_neighbors_I = 1- number_of_neighbors_I./number_of_neighbors;
     
-    system.vaccinated(indices_not_vacc) = not(system.vaccinated(indices_not_vacc));
+    % Uniform distribition to change your state to vaccinated or not
+    proba_change_state = rand(n);
+    
+    % Probability to change the vaccination choice
+%     size(rewards_of_neighbors_V)
+%     size(number_of_neighbors_V)
+%     size(rewards_of_neighbors_NV)
+%     size(number_of_neighbors_NV)
+%     (rewards_of_neighbors_V-rewards_of_neighbors_NV).*number_of_neighbors_V ...
+%                             + (rewards_of_neighbors_NV-rewards_of_neighbors_V).*number_of_neighbors_NV
+    proba_vaccination = vaccination_probability_1((-rewards_of_neighbors_V+rewards_of_neighbors_NV).*number_of_neighbors_V ...
+                            + (-rewards_of_neighbors_NV+rewards_of_neighbors_V).*number_of_neighbors_NV);

```

Comment:
Also, suppose two systems:
<pre>
S1 : reward | n° of people | choice            S2 : reward | n° of people | choice
     -10    |       6      |   Vacc                  -10   |       6      |  Non Vacc
     -2     |       3      |  Non Vacc                -2   |       3      |    Vacc
</pre>
The quantity (-r_V+r_NV)*n°_V + (-r_NV + r_V)*n°_NV is equal to
(10-2)*6 + (2-10)*3 = 8 * 6 - 8 * 3 = 24 for S1
(2-10)*3 + (10-2)*6 = -8 * 3 + 8 * 6 = 24 for S2

So how does this function differentiate these two cases?
For S1 the person should choose to be Non Vacc, and for S2 the person should choose to be Vacc, but the function will return the same number for both cases.

Reply:
You’re right it’s a mistake “n°_V” should be replaced by “system.vaccinated” and “n°_NV” by “~system.vaccinated”

Reply:
ok, that makes more sense. But then are you sure about the shape of the function with delta = 0.99? I might be wrong but I think it would be more correct to write (r_V - r_NV).*system.vacc + (r_NV - r_V).*~system_vacc and put delta = 0.01. Maybe plot the function on wolfram alpha to make sure it does what you want it to

Reply:
Maybe, I will check that later today ;)

Reply:
My bad, it actually works as youve written it because of the way you've written the inequality 
proba_vacc < proba_change_state

Reply:
so all good for me